Cargo Workspaces and Partial Config defaults

dbanty opened the pull request:
See the `.changeset` files for more detail, but basically, this should allow you to use a workspace `Cargo.toml` file with fixed members instead of manually configuring multiple packages in `knope.toml`.

TODO:

- [x] Run through a test of the updated monorepo tutorial
- [x] Add default `scopes` for each package matching the package's name

FallenValkyrie reviewed (changes requested):
I really love the addition of partial configuration. I do have a few neat things we could include in this PR and a few nits.

I'll have another look tomorrow, I kinda lost concentration half way, so i may have missed a few things

FallenValkyrie commented on .changeset/add_basic_cargo_workspace_support.md:
````diff
@@ -0,0 +1,19 @@
+]
+```
+
+then Knope will now treat each member like a package, assuming that there is a `Cargo.toml` file for each member.
````
a note on this: 
I think it is always better to be explicit, rather than implicit when it comes to user interfaces, unless the interface itself tells you when an assumption it made turned out to be wrong.

dbanty replied:
Yes, good point. According to the Cargo workspace docs, it's an error to point at a member which does not have a Cargo.toml file, so I just made it an error as well.

FallenValkyrie commented:
> Run through a test of the updated monorepo tutorial

I'll also test this tomorrow and see if anything is unclear in the process ^^

dbanty commented on docs/src/content/docs/tutorials/releasing-multiple-packages.mdx:
````diff
@@ -46,30 +46,17 @@ gh repo create --private knope-tutorial --source .
+Instead, create this by hand:
 
-```toml title="knope.toml" ins={2-3,6-7,10-11}
+```toml title="knope.toml" {"Replace this with your GitHub or Gitea info": 12-15}
````
It's a new feature!

dbanty commented on src/config/mod.rs:
```diff
@@ -308,21 +337,37 @@ pub(crate) fn generate() -> Config {
+fn generate_workflows(has_forge: bool, packages: &[Package]) -> Vec<Workflow> {
+    let (commit_message, variables) = if packages.len() < 2 {
+        let mut variables = IndexMap::new();
+        variables.insert(String::from("$version"), Variable::Version);
```
Are you suggesting that we default to replacing the string `$version` with the version variable, even if that variable wasn't declared? That would be tricky because someone might want that literal string and now doesn't have an obvious way to opt out. Or maybe I'm misunderstanding...

FallenValkyrie replied:
mmh we could give the user a simple flag `defaults` field on `variables::Template` they can set to false per step where default behavior isn't wished and no other variables are set, if all they care for is the literal `$version`. hat way they can override or disable the default behavior. Additionally i would only default to this if no variables are configured at all.
 
 At least from my perspective when using it, i pretty much exclusively used the variables as described in the docs. Mainly because i think it's a sane default ^^

dbanty commented on src/step/releases/workspace.rs:
```diff
@@ -0,0 +1,64 @@
+            .get("members")?
+            .as_array()?
+            .iter()
+            .filter_map(|member| member.as_str())
```
Literally only because `.as_str()` returns an `Option`... it _should_ never return `None` unless someone has something other than a string in the members array. We could error instead of ignoring those cases though...

FallenValkyrie replied:
I think an error would indeed make more sense. I mean a number makes no sense in that position :p

dbanty commented on tests/bump_version.rs:
```diff
@@ -8,7 +8,7 @@ use snapbox::cmd::{cargo_bin, Command};
 mod helpers;
 
-/// Test all the `BumpVersion` rules.
+/// Test all the `Bumname = "default"` rules.
```
Ah, bad search & replace from trying to update all the test `Cargo.toml` files everywhere 😅

dbanty commented:
@FallenValkyrie thanks so much for the review! Sorry it's such a big PR 😓

FallenValkyrie commented:
> @FallenValkyrie thanks so much for the review! Sorry it's such a big PR 😓

hah, don't worry, i'm actually quite excited. ADHD just makes shit hard, so don't worry too much ^^

FallenValkyrie commented on src/step/releases/workspace.rs:
```diff
@@ -0,0 +1,88 @@
+    let workspace_path = path.parent().ok_or_else(|| Error::Parent(path.into()))?;
+    let Some(members) = cargo_toml
+        .get("workspace")
+        .and_then(|workspace| workspace.as_table())
+        .and_then(|workspace| workspace.get("members"))
+        .and_then(|members| members.as_array())
```
a little fun fact: 
the `?` operator also works for `Option`, so this could be:
```suggestion
        .and_then(|workspace| {
             workspace
                 .as_table()?
                 .get("members")?
                 .as_array()
         })
```

it's an absolute nit and up to you what you find nicer ^^

dbanty replied:
I do like that a little better, reads as more performant, too (even though `rustc` probably optimizes away the extra closures)

FallenValkyrie reviewed (changes requested):
Here we go, Just see this as a continuation of the other one ^^

dbanty commented:
I think I addressed all the suggestions except the default variable thing, I could probably be convinced in a follow-up issue, but I'm a really big fan of explicit config and the current method makes it pretty clear when a variable is supported in a location (I think).

dbanty commented:
Tutorial updated! (It needed a little love)

FallenValkyrie commented:
> I think I addressed all the suggestions except the default variable thing, I could probably be convinced in a follow-up issue, but I'm a really big fan of explicit config and the current method makes it pretty clear when a variable is supported in a location (I think).

Makes sense to me ^^

FallenValkyrie reviewed (approved):
I cannot find anything worth mentioning, so LGTM. 🚀
